src/sniffer.c

```diff
@@ -2155,701 +2189,861 @@ static void ShowTlsSecrets(SnifferSession* session)
+                ret = wc_ecc_init_ex(&args->key->priv.ecc, NULL, devId);
+                if (ret == 0) {
+                    args->key->type = WC_PK_TYPE_ECDH;
+                    args->key->initPriv = 1;
+                #ifdef WOLFSSL_ASYNC_CRYPT
+                    ret = wolfSSL_AsyncInit(ssl, &args->key->priv.ecc.asyncDev,
                         WC_ASYNC_FLAG_CALL_AGAIN);
-            #endif
-                if (ret >= 0) {
-                    PRIVATE_KEY_UNLOCK();
-                    ret = wc_DhAgree(&dhKey,
-                        session->sslServer->arrays->preMasterSecret,
-                        &session->sslServer->arrays->preMasterSz,
-                        privKey, privKeySz,
-                        input, *sslBytes);
-                    PRIVATE_KEY_LOCK();
+                #endif
                 }
-            } while (ret == WC_PENDING_E);
         }
+        #if defined(ECC_TIMING_RESISTANT) && (!defined(HAVE_FIPS) || \
+            (!defined(HAVE_FIPS_VERSION) || (HAVE_FIPS_VERSION != 2))) && \
+            !defined(HAVE_SELFTEST)
+            if (ret == 0) {
+                ret = wc_ecc_set_rng(&args->key->priv.ecc, session->sslServer->rng);
+            }
+        #endif
+            if (ret == 0) {
+                idx = 0;
+                ret = wc_EccPrivateKeyDecode(args->keyBuf->buffer, &idx,
+                    &args->key->priv.ecc, args->keyBuf->length);
+                if (ret != 0) {
+                    SetError(ECC_DECODE_STR, error, session, FATAL_ERROR_STATE);
+                }
+            }
 
-        if (keyInit)
-            wc_FreeDhKey(&dhKey);
+            if (ret == 0) {
+                args->length = wc_ecc_size(&args->key->priv.ecc) * 2 + 1;
+                /* The length should be 2 times the key size (x and y), plus 1
+                * for the type byte. */
+                if (!IsAtLeastTLSv1_3(session->sslServer->version)) {
+                    input += 1; /* Don't include the TLS length for the key. */
+                }
 
-    #ifdef WOLFSSL_SNIFFER_STATS
-        if (ret != 0)
-            INC_STAT(SnifferStats.sslKeyFails);
-    #endif
+                if (args->length > *sslBytes) {
+                    SetError(PARTIAL_INPUT_STR, error, session, FATAL_ERROR_STATE);
+                    ret = -1;
+                }
 
-        /* left-padded with zeros up to the size of the prime */
-        if (ret == 0 && p_len > session->sslServer->arrays->preMasterSz) {
-            word32 diff = p_len - session->sslServer->arrays->preMasterSz;
-            XMEMMOVE(session->sslServer->arrays->preMasterSecret + diff,
-                    session->sslServer->arrays->preMasterSecret,
-                    session->sslServer->arrays->preMasterSz);
-            XMEMSET(session->sslServer->arrays->preMasterSecret, 0, diff);
-            session->sslServer->arrays->preMasterSz = p_len;
+                /* if curve not provided in key share data, then use private key curve */
```

Comment:
If curve was provided in key share, should we do a consistency check against the private key curve?  (or is that done somewhere else?)

Reply:
It is done in `wc_ecc_import_x963_ex` below.